language/bytecode-verifier/src/abstract_state.rs

```diff
@@ -129,452 +110,192 @@ impl AbstractState {
-        self.partition.remove_nonce(nonce);
+    /// gets all borrows of `nonce` that might be consistent with a field
+    pub fn field_borrows(&self, nonce: Nonce, idx: FieldDefinitionIndex) -> BTreeSet<Nonce> {
+        self.borrow_graph
+            .consistent_borrows(nonce, LabelElem::Field(idx))
     }
 
-    /// checks if there are any pending borrows on value
-    pub fn is_full(&self, value: &AbstractValue) -> bool {
-        match value {
-            AbstractValue::Reference(nonce) => !self.borrows.contains_key(&nonce),
-            AbstractValue::Value(_, nonce_set) => nonce_set.is_empty(),
-        }
+    /// checks if `nonce` is borrowed
+    pub fn is_nonce_borrowed(&self, nonce: Nonce) -> bool {
+        !self.borrow_graph.all_borrows(nonce).is_empty()
     }
 
-    /// returns the set of nonces borrowing from nonce that might alias some idx-extension of nonce
-    pub fn borrowed_nonces_for_field(
-        &self,
-        idx: FieldDefinitionIndex,
-        nonce: Nonce,
-    ) -> BTreeSet<Nonce> {
-        if self.borrows.contains_key(&nonce) {
-            match &self.borrows[&nonce] {
-                BorrowInfo::BorrowedBy(borrowed_by) => borrowed_by.clone(),
-                BorrowInfo::FieldsBorrowedBy(fields_borrowed_by) => {
-                    if fields_borrowed_by.contains_key(&idx) {
-                        fields_borrowed_by[&idx].clone()
-                    } else {
-                        BTreeSet::new()
-                    }
-                }
-            }
-        } else {
-            BTreeSet::new()
-        }
+    /// checks if `nonce` is borrowed via an edge with the empty label
+    pub fn is_nonce_nil_borrowed(&self, nonce: Nonce) -> bool {
+        !self.borrow_graph.nil_borrows(nonce).is_empty()
     }
 
-    /// returns the set of nonces borrowing from nonce that might alias some extension of nonce
-    pub fn borrowed_nonces(&self, nonce: Nonce) -> BTreeSet<Nonce> {
-        if self.borrows.contains_key(&nonce) {
-            match &self.borrows[&nonce] {
-                BorrowInfo::BorrowedBy(borrowed_by) => borrowed_by.clone(),
-                BorrowInfo::FieldsBorrowedBy(fields_borrowed_by) => {
-                    let empty_set = BTreeSet::new();
-                    Self::get_union_of_sets(&empty_set, fields_borrowed_by)
-                }
-            }
-        } else {
-            BTreeSet::new()
-        }
+    /// checks if a local is borrowed
+    pub fn is_local_borrowed(&self, idx: LocalIndex) -> bool {
+        !self.local_borrows(idx).is_empty()
     }
 
-    /// update self to reflect a borrow of idx from nonce by new_nonce
+    /// checks if a global is borrowed
+    pub fn is_global_borrowed(&self, idx: StructDefinitionIndex) -> bool {
+        !self.global_borrows(idx).is_empty()
+    }
+
+    /// update self to reflect a borrow of a value global@idx by new_nonce
+    pub fn borrow_from_global_value(&mut self, idx: StructDefinitionIndex, new_nonce: Nonce) {
```

Comment:
Are you borrowing a global value, or are you borrowing something from a global value? I suspect the former, the name suggest the latter.

Reply:
I agree. I will rename to borrow_global_value and borrow_local_value, respectively.